<h1>Exploring Hacker News Posts</h1>

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set here, but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

    id: The unique identifier from Hacker News for the post
    title: The title of the post
    url: The URL that the posts links to, if it the post has a URL
    num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
    num_comments: The number of comments that were made on the post
    author: The username of the person who submitted the post
    created_at: The date and time at which the post was submitted

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

    Ask HN: How to improve my personal website?

    Ask HN: Am I the only one outraged by Twitter shutting down share counts?

    Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

    Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'

    Show HN: Something pointless I made

    Show HN: Shanhu.io, a programming playground powered by e8vm

We'll compare these two types of posts to determine the following:

    Do Ask HN or Show HN receive more comments on average?
    Do posts created at a certain time receive more comments on average?

Let's start by importing the libraries we need and reading the data set into a list of lists.

In [1]:
from csv import reader
opened_file = open("E:/HN_posts_year_to_Sep_26_2016.csv", encoding = "cp437")
read_file = reader(opened_file)
hn = list(read_file)
loop = 0
while (loop < 5):
    print(hn[loop])
    loop += 1

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26']
['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24']
['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19']
['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']


In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)
loop = 0
while (loop < 5):
    print(hn[loop])
    loop += 1

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26']
['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24']
['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19']
['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']
['12578979', 'How the Data Vault Enables the Next-Gen Data Warehouse and Data Lake', 'https://www.talend.com/blog/2016/05/12/talend-and-“the-data-vault”', '1', '0', 'markgainor1', '9/26/2016 3:14']


In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1].lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

9139
10158
273822


In [4]:
total_ask_comments = 0
for ask in ask_posts:
    total_ask_comments += int(ask[4])
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)
total_show_comments = 0
for ask in show_posts:
    total_show_comments += int(ask[4])
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.393478498741656
4.886099625910612


<h2>Best Hours to Make an Ask Posts</h2>

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

    Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
    Calculate the average number of comments ask posts receive by hour created.

In [5]:
import datetime as dt
result_list = []
for ask in ask_posts:
    create = ask[6]
    comment = int(ask[4])
    result_list.append([create, comment])
    
counts_by_hour = {}
comments_by_hour = {}
for lists in result_list:
    hour = lists[0]
    hour_dt = dt.datetime.strptime(hour, "%m/%d/%Y %H:%M")
    hour_str = hour_dt.strftime("%H")
    if hour_str not in counts_by_hour:
        counts_by_hour[hour_str] = 1
        comments_by_hour[hour_str] = lists[1]
    else:
        counts_by_hour[hour_str] += 1
        comments_by_hour[hour_str] += lists[1]

In [14]:
avg_by_hour = []
for lists in comments_by_hour:
    avg_comment = comments_by_hour[lists] / counts_by_hour[lists]
    avg_by_hour.append([lists, avg_comment])

In [15]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Posts Comments")
for loop in sorted_swap[:5]:
    swap_hour = loop[1]
    swap_hour_dt = dt.datetime.strptime(swap_hour, "%H")
    swap_hour_new = swap_hour_dt + dt.timedelta(hours = 13)
    swap_hour_str = swap_hour_new.strftime("%H")
    swap_avg = loop[0]
    string = "{}:00: {:.2f} average comments per post."
    print(string.format(swap_hour_str, swap_avg))

Top 5 Hours for Ask Posts Comments
04:00: 28.68 average comments per post.
02:00: 16.32 average comments per post.
01:00: 12.38 average comments per post.
15:00: 11.14 average comments per post.
23:00: 10.68 average comments per post.


The hour that receives the most comments per post on average is at 4 AM with 28.68 average comments per post, which is 75.73% increase from the second highest average comments per post.

<h2>Best Hours to Make a Show Posts</h2>

Next, we'll determine if show posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

    Calculate the amount of show posts created in each hour of the day, along with the number of comments received.
    Calculate the average number of comments show posts receive by hour created.

In [17]:
result_list = []
for show in show_posts:
    create = show[6]
    comment = int(show[4])
    result_list.append([create, comment])
    
counts_by_hour = {}
comments_by_hour = {}
for lists in result_list:
    hour = lists[0]
    hour_dt = dt.datetime.strptime(hour, "%m/%d/%Y %H:%M")
    hour_str = hour_dt.strftime("%H")
    if hour_str not in counts_by_hour:
        counts_by_hour[hour_str] = 1
        comments_by_hour[hour_str] = lists[1]
    else:
        counts_by_hour[hour_str] += 1
        comments_by_hour[hour_str] += lists[1]

In [18]:
avg_by_hour = []
for lists in comments_by_hour:
    avg_comment = comments_by_hour[lists] / counts_by_hour[lists]
    avg_by_hour.append([lists, avg_comment])

In [20]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Show Posts Comments")
for loop in sorted_swap[:5]:
    swap_hour = loop[1]
    swap_hour_dt = dt.datetime.strptime(swap_hour, "%H")
    swap_hour_new = swap_hour_dt + dt.timedelta(hours = 13)
    swap_hour_str = swap_hour_new.strftime("%H")
    swap_avg = loop[0]
    string = "{}:00: {:.2f} average comments per post."
    print(string.format(swap_hour_str, swap_avg))

Top 5 Hours for Show Posts Comments
01:00: 6.99 average comments per post.
20:00: 6.68 average comments per post.
00:00: 6.00 average comments per post.
21:00: 5.60 average comments per post.
03:00: 5.52 average comments per post.


The hour that receives the most comments per post on average is at 1 AM with average 6.99 comments per post, with only 4.4% increase from the second highest average comments from the post.

<h1>Conclusion</h1>

The best hour to create a post on Hacker News is around 1-4 AM (GMT +8) because you can get a lot of comments and responds.